res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,718 @@
+        engine.scratchTick(deckNum, newVal);
+    }
+    else{ // fallback
+        engine.setValue(group, 'jog', newVal * this.bendScale);
+    }
+
+};
+
+
+PioneerDDJ400.jogSearch = function(_channel, _control, value, _status, group) {
+    // "highspeed" (scaleup value) pitch bend
+    const newVal = (value - 64) * this.highspeedScale;
+    engine.setValue(group, 'jog', newVal);
+};
+
+PioneerDDJ400.jogTouch = function(channel, _control, value) {
+    const deckNum = channel + 1;   
+
+    // skip scratchmode if we adjust the loop points
+    if(this.loopin4beat[channel] || this.loopout[channel]){
+        return;
+    }
+
+    // on touch jog with vinylmode enabled -> enable scratchmode
+    if(value != 0 && this.vinylMode){
+        engine.scratchEnable(deckNum, 720, 33+1/3, this.alpha, this.beta);
+    }
+    else{
+        // on release jog (value==0) disable pitch bend mode or scratch mode
+        engine.scratchDisable(deckNum);
+    }
+};
+
+PioneerDDJ400.cycleTempoRange = function(_channel, _control, value, _status, group){
+    if(value == 0) return; // ignore release
+    
+    const currRange = engine.getValue(group, 'rateRange');
+    var idx = 0;
+
+    for(var i = 0; i < this.tempoRanges.length; i++){
+        if(currRange == this.tempoRanges[i]){
+            idx = (i + 1) % this.tempoRanges.length;
+            break;
+        }
+    }
+    engine.setValue(group, 'rateRange', this.tempoRanges[idx]);
+};
+
+
+function sortAsc(a, b){
+    return a > b ? 1 : b > a ? -1 : 0; 
+}
+
+PioneerDDJ400.initCuePointsAndLoops = function(group){
+    // create a list of positions in the track which can be selected
+    var points = [];
+
+    for(var padNum = 1; padNum <= 8; padNum++){
+        points.push( engine.getValue(group, 'hotcue_'+padNum+'_position'));
+    }
+    points.push( engine.getValue(group, 'cue_point'));
+    points.push( engine.getValue(group, 'loop_start_position'));
+    points.push( engine.getValue(group, 'loop_end_position'));
+    points.sort(sortAsc); // sort asc
+
+    return points;
+};
+
+PioneerDDJ400.cueLoopCallLeft = function(_channel, _control, value, _status, group){
+    if(value == 0) return; // ignore release
+
+    const loop_on = engine.getValue(group, 'loop_enabled');
+    if (loop_on){
+        // loop halve
+        engine.setValue(group, 'loop_scale', 0.5);
+    }
+    else{
+        const currentPosition = engine.getValue(group, 'playposition') - this.pointJumpSpace;
+        const trackSamples = engine.getValue(group, 'track_samples');
+        const points = this.initCuePointsAndLoops(group);
+        var newPosition = currentPosition;
+
+        for(var i = 1; i <= points.length; i++){
+            if(i == points.length || points[i] >= currentPosition * trackSamples){
+                newPosition = points[i-1] / trackSamples;
+                break;
+            }
+        }
+        //engine.setValue(group, 'loop_in_goto', 1);
+        engine.setValue(group, 'playposition', newPosition);
```

Comment:
I do not understand the purpose of this. This controller has hotcue buttons, so what does this add?